### DISCUSSION TITLE
About hybrid search workflow

### DISCUSSION THREAD

**Discussion by ruru5697** [milvus-io/milvus#23944] [Q&A and General discussion] [answered]
Due to business needs, I need to create a feature library with more than 100,000 partitions. Since milvus has a limit on the number of collections and partitions, I built 5,000 collections, each containing 20 partitions.
However, with the increase of the number of collections and partitions, the milvus log grows very fast, and each collection will execute the load method immediately after it is created.
When the number of partitions exceeds 60,000, the milvus server will not respond, and new collections and partitions can no longer be created.

To alleviate this problem, only one collection and one partition are now created, and the tag field is defined in the feature library to manage feature sharding.

```
primary_key = FieldSchema(name='primary_key', dtype=DataType.INT64, is_primary=True, auto_id=auto_id)
feature = FieldSchema(name='feature', dtype=DataType.FLOAT_VECTOR, dim=feature_dim)
tag = FieldSchema(name='tag', dtype=DataType.VARCHAR, max_length=extra_data_length)
schema = CollectionSchema([primary_key, feature, extra_data], 
                          description=json.dumps({'replica_number': collection_config['replica_number']}))
collection = Collection(
    name=collection_config['collection_name'],
    schema=schema,
    shards_num=collection_config['shards_num']
)

index_params = {
    'index_type': 'IVF_FLAT',
    'metric_type': 'IP',
    'params': {'nlist': 100}
}
collection.create_index(field_name='feature', index_params=index_params)
```

However, after completing feature insertion, there was a problem retrieving features.

```
c.search(data=feature.tolist(), anns_field='feature', param={'metric_type': "IP", 'params': {'nprobe': 32}}, output_fields=['tag'], limit=1, consistency_level='Strong', expr='tag == "21345"')
```

When nprobe < nlist, the features under the corresponding tag may not be recalled, and only when nprobe == nlist, the features under the corresponding tag can be recalled.

What is the search process of hybrid search, when is the conditional search performed? Can feature search perform after the conditional search of tag?

Or how to design the partition structure of milvus under the need for a large number of feature partitions.

**Comment by yhmo** [comment]
In milvus, each collection will subscribe messages from pulsar/kafka, and each collection will send time tick message to pulsar/kafka per 200ms. With more and more collections created, large amounts of time tick actions will slow down the entire system and run into disaster.
It is recommended to reduce collections/partitions number to a small value. Sometimes we use hash method to distribute data into different collections or partitions.

**Reply by ruru5697** [reply]
Is it recommended to use additional fields as partition identifiers in a collection/partition?

**Reply by yhmo** [reply]
> Is it recommended to use additional fields as partition identifiers in a collection/partition?

It depends on your requirement.
Assume your application has 1000 users, you create 1 collection, and 10 partitions in the collection.
You hash user1, user11, user21 .... into partition1, hash user2, user12, user22 .... into partition2, etc. Then you can do hybrid search like: `search(expr="user_id in [user11, user21, user22]", partition_names=[partition1, partition2])`
But sometimes you want more filtering conditions, such as "the user is female or male?", then you need to define an additional field "gender" for the collection. So that you can do hybrid search like:
`search(expr="gender==1", partition_names=[partition1, partition2])`

**Comment by yhmo** [comment, marked as answer]
"When nprobe < nlist, the features under the corresponding tag may not be recalled, and only when nprobe == nlist, the features under the corresponding tag can be recalled."

For hybrid search, milvus first scans all the rows, and marks some rows to be "skip" if it is filtered out by the expression. Then pick nprobe buckets to do ANN search.
Assume there are nlist=100 buckets, you call search with nprobe=10, and the `expr='tag == "21345"'`. Unfortunately, all the rows in the 10 buckets are filtered out, the search result will be empty. That is why you saw some items may not be recalled.

**Reply by ruru5697** [reply]
Do your team have plans to support feature search on expression filtered results？This may improve the reliability of hybrid search results.

**Reply by yhmo** [reply]
We don't have this plan.
Filtering after ANN search, it always computes distance for all the vectors in nprobe bucket. Filtering before ANN search, some vectors are skipped and save the compute time.

**Reply by winterfellding** [reply]
Hi, @yhmo the nprobe bucket here, does it mean the n nestest centroid bucket to the query vector or the bucket list is generated from the candidate entity id filtered from expr

**Comment by xiaofan-luan** [comment]
> Due to business needs, I need to create a feature library with more than 100,000 partitions. Since milvus has a limit on the number of collections and partitions, I built 5,000 collections, each containing 20 partitions. However, with the increase of the number of collections and partitions, the milvus log grows very fast, and each collection will execute the load method immediately after it is created. When the number of partitions exceeds 60,000, the milvus server will not respond, and new collections and partitions can no longer be created.
> 
> To alleviate this problem, only one collection and one partition are now created, and the tag field is defined in the feature library to manage feature sharding.
> 
> ```
> primary_key = FieldSchema(name='primary_key', dtype=DataType.INT64, is_primary=True, auto_id=auto_id)
> feature = FieldSchema(name='feature', dtype=DataType.FLOAT_VECTOR, dim=feature_dim)
> tag = FieldSchema(name='tag', dtype=DataType.VARCHAR, max_length=extra_data_length)
> schema = CollectionSchema([primary_key, feature, extra_data], 
>                           description=json.dumps({'replica_number': collection_config['replica_number']}))
> collection = Collection(
>     name=collection_config['collection_name'],
>     schema=schema,
>     shards_num=collection_config['shards_num']
> )
> 
> index_params = {
>     'index_type': 'IVF_FLAT',
>     'metric_type': 'IP',
>     'params': {'nlist': 100}
> }
> collection.create_index(field_name='feature', index_params=index_params)
> ```
> 
> However, after completing feature insertion, there was a problem retrieving features.
> 
> ```
> c.search(data=feature.tolist(), anns_field='feature', param={'metric_type': "IP", 'params': {'nprobe': 32}}, output_fields=['tag'], limit=1, consistency_level='Strong', expr='tag == "21345"')
> ```
> 
> When nprobe < nlist, the features under the corresponding tag may not be recalled, and only when nprobe == nlist, the features under the corresponding tag can be recalled.
> 
> What is the search process of hybrid search, when is the conditional search performed? Can feature search perform after the conditional search of tag?
> 
> Or how to design the partition structure of milvus under the need for a large number of feature partitions.



> Due to business needs, I need to create a feature library with more than 100,000 partitions. Since milvus has a limit on the number of collections and partitions, I built 5,000 collections, each containing 20 partitions. However, with the increase of the number of collections and partitions, the milvus log grows very fast, and each collection will execute the load method immediately after it is created. When the number of partitions exceeds 60,000, the milvus server will not respond, and new collections and partitions can no longer be created.
> 
> To alleviate this problem, only one collection and one partition are now created, and the tag field is defined in the feature library to manage feature sharding.
> 
> ```
> primary_key = FieldSchema(name='primary_key', dtype=DataType.INT64, is_primary=True, auto_id=auto_id)
> feature = FieldSchema(name='feature', dtype=DataType.FLOAT_VECTOR, dim=feature_dim)
> tag = FieldSchema(name='tag', dtype=DataType.VARCHAR, max_length=extra_data_length)
> schema = CollectionSchema([primary_key, feature, extra_data], 
>                           description=json.dumps({'replica_number': collection_config['replica_number']}))
> collection = Collection(
>     name=collection_config['collection_name'],
>     schema=schema,
>     shards_num=collection_config['shards_num']
> )
> 
> index_params = {
>     'index_type': 'IVF_FLAT',
>     'metric_type': 'IP',
>     'params': {'nlist': 100}
> }
> collection.create_index(field_name='feature', index_params=index_params)
> ```
> 
> However, after completing feature insertion, there was a problem retrieving features.
> 
> ```
> c.search(data=feature.tolist(), anns_field='feature', param={'metric_type': "IP", 'params': {'nprobe': 32}}, output_fields=['tag'], limit=1, consistency_level='Strong', expr='tag == "21345"')
> ```
> 
> When nprobe < nlist, the features under the corresponding tag may not be recalled, and only when nprobe == nlist, the features under the corresponding tag can be recalled.
> 
> What is the search process of hybrid search, when is the conditional search performed? Can feature search perform after the conditional search of tag?
> 
> Or how to design the partition structure of milvus under the need for a large number of feature partitions.

use the new partition key should solve the issue.

What we do is:
first parse your filtering expression and map your key to physical partition
filter based on the filtering expression
based on the filtering result, decide whether to brute force search or use index for topk result

**Reply by ruru5697** [reply]
Does this mean that I can use the partition key under 1 collection and 1 partition to break through the limit of partition in 1 collection(4096), and avoiding this issue

> In milvus, each collection will subscribe messages from pulsar/kafka, and each collection will send time tick message to pulsar/kafka per 200ms. With more and more collections created, large amounts of time tick actions will slow down the entire system and run into disaster. It is recommended to reduce collections/partitions number to a small value. Sometimes we use hash method to distribute data into different collections or partitions.

If I have to create 10,000 parition and every partition include 30~50 features,  Is partition key a better solution?

**Reply by xiaofan-luan** [reply]
it's not gonna to be perfect. But if each partition contains only 30-50 features. why not simply store each partition in a redis or databases.

**Reply by xiaofan-luan** [reply]
ann doesn't make any sense on data less than 10K

**Comment by xiaofan-luan** [comment]
My suggestion is node to create too much collections. partition key is designed for that.
